### PR TITLE
Updating logic to check if a list of observables is qwc

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
Updating the logic which checks if a list of observables (assumed to be Pauli-words) is qubit wise commuting

**Description of the Change:**
Either adding a function or upgrading the logic present to reduce the runtime of the qwc check. 

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
- [Feature request](https://github.com/PennyLaneAI/pennylane/issues/2788)
- [Issue thread](https://github.com/PennyLaneAI/pennylane/issues/2706)

### CODE REVIEW

**Comment by Jaybsoni** [comment]
All that's left is to add some tests,

**Comment by Jaybsoni** [comment]
Here are some benchmarking plots which compare the performance the old vs. new logic for checking qwc of a list of pauli words. 

image

Looking at the relative scale we are seeing between a 25x to 500x improvement in execution time. 

In both graphs we see a constant, almost instantaneous execution time for the observable containing X,Z terms that is not qubit wise commuting (purple). This is due to early termination of the program once its comes across an incompatibility between the observables. 

In the bottom graph we see that the observable with only Z terms (red) is also very performant. This is because of dedicated logic in the previous implementation which checked for the special case when the observable was made up of only one type of operator. Below I provide graphs for each observable type individually: 

image
image

Here we see performance improvements for each of the observable types. There is a performance regression for the all Z case, but its on the order of 0.005 sec and is thus negligible.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -82,9 +82,7 @@
   [(#2721)](https://github.com/PennyLaneAI/pennylane/pull/2721)
 
 * Added operation `qml.QutritUnitary` for applying user-specified unitary operations on qutrit devices.
-  [(#2699)](https://github.com/PennyLaneAI/pennylane/pull/2699)  
-
-**Operator Arithmetic:**
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:87]: There were two **Operator Arithmetic** headers, I just removed one

File: doc/releases/changelog-dev.md [hunk 2]
```diff
@@ -171,7 +169,8 @@
   labels.
   [(#2779)](https://github.com/PennyLaneAI/pennylane/pull/2779)
 
-* Adds a new function to compare operators. `qml.equal` can be used to compare equality of parametric operators taking into account their interfaces and trainability.
+* Adds a new function to compare operators. `qml.equal` can be used to compare equality of parametric operators taking 
+  into account their interfaces and trainability.
```
Comment (Jaybsoni) [inline doc/releases/changelog-dev.md:173]: This was spilling over multiple lines, so cleaned it up

File: pennylane/grouping/utils.py [hunk 3]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+    Returns:
+        (bool): True if they are all qubit-wise commuting, false otherwise.
+    """
+    wire_dict = {}
+
+    for op in lst_pauli_words:  # iterate over the list of observables
+        op_names = [op.name] if not isinstance(op.name, list) else op.name
+        op_wires = op.wires.tolist()
+
+        for name, wire in zip(op_names, op_wires):
```
Comment (Jaybsoni) [inline pennylane/grouping/utils.py:679]:
```suggestion
        for name, wire in zip(op_names, op_wires):  # iterate over wires, this is at most n 
```

**Comment by Jaybsoni** [comment]
[sc-22483]

File: pennylane/grouping/utils.py [hunk 4]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+    Returns:
+        (bool): True if they are all qubit-wise commuting, false otherwise.
+    """
+    wire_dict = {}
+
+    for op in lst_pauli_words:  # iterate over the list of  m observables 
+        op_names = [op.name] if not isinstance(op.name, list) else op.name
+        op_wires = op.wires.tolist()
+
+        for name, wire in zip(op_names, op_wires):  # iterate over wires, this is at most n
```
Comment (Jaybsoni) [inline pennylane/grouping/utils.py:679]:
```suggestion
        for name, wire in zip(op_names, op_wires):  # iterate over wires, this is at most n
```

File: pennylane/grouping/utils.py [hunk 5]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+
+    Args:
+        lst_pauliwords (list[Observable]): list of observables (assumed to be valid Pauli words).
+
+    Returns:
+        (bool): True if they are all qubit-wise commuting, false otherwise.
+    """
+    wire_dict = {}
+
+    for op in lst_pauli_words:  # iterate over the list of  m observables
```
Comment (Jaybsoni) [inline pennylane/grouping/utils.py:675]:
```suggestion
    for op in lst_pauli_words:  # iterate over the list of  m observables
```

File: pennylane/grouping/utils.py [hunk 6]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
     return True
 
 
+def are_pauli_words_qwc(lst_pauli_words):
+    """Given a list of observables assumed to be valid Pauli words, determine if they
+     are all mutually qubit-wise commuting.
+
+    This implementation has time complexity ~ O(m * n) for n wires and m Pauli words.
```
Comment (antalszava) [inline pennylane/grouping/utils.py:665]: What does `n wires` mean? No wires seem to have been mentioned/defined up until here.

File: pennylane/grouping/utils.py [hunk 7]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+
+    Args:
+        lst_pauliwords (list[Observable]): list of observables (assumed to be valid Pauli words).
+
+    Returns:
+        (bool): True if they are all qubit-wise commuting, false otherwise.
+    """
+    wire_dict = {}
+
+    for op in lst_pauli_words:  # iterate over the list of  m observables
```
Comment (antalszava) [inline pennylane/grouping/utils.py:675]: (Minor): m seems a bit confusing when just looking at the code.
```suggestion
    for op in lst_pauli_words:  # iterate over the list of observables
```

File: pennylane/grouping/utils.py [hunk 8]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+        op_names = [op.name] if not isinstance(op.name, list) else op.name
+        op_wires = op.wires.tolist()
+
+        for name, wire in zip(op_names, op_wires):  # iterate over wires, this is at most n
```
Comment (antalszava) [inline pennylane/grouping/utils.py:679]: What is n? Need an explanation here too.

File: pennylane/grouping/utils.py [hunk 9]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
+
+    This implementation has time complexity ~ O(m * n) for n wires and m Pauli words.
+
+    Args:
+        lst_pauliwords (list[Observable]): list of observables (assumed to be valid Pauli words).
+
+    Returns:
+        (bool): True if they are all qubit-wise commuting, false otherwise.
+    """
+    wire_dict = {}
```
Comment (antalszava) [inline pennylane/grouping/utils.py:673]: Worth making this name more descriptive or adding a comment on what the object is used for.

E.g., later when reading `wire_dict[wire]`, the reader should be able to tell what the value of that expression is. Right now it's not apparent.
```suggestion
    latest_op_name_per_wire = {}
```

File: pennylane/grouping/utils.py (hunk 8 above)
Comment (antalszava) [inline pennylane/grouping/utils.py:679]: `name` -> `op_name` renaming could make the code more readable I think.

File: pennylane/grouping/utils.py [hunk 10]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
...
 
 
+def are_pauli_words_qwc(lst_pauli_words):
+    """Given a list of observables assumed to be valid Pauli words, determine if they
+     are all mutually qubit-wise commuting.
+
+    This implementation has time complexity ~ O(m * n) for n wires and m Pauli words.
+
+    Args:
+        lst_pauliwords (list[Observable]): list of observables (assumed to be valid Pauli words).
```
Comment (antalszava) [inline pennylane/grouping/utils.py:668]:
```suggestion
        lst_pauliwords (list[Observable]): List of observables (assumed to be valid Pauli words).
```

File: tests/grouping/test_grouping_utils.py [hunk 11]
```diff
@@ -195,6 +196,26 @@ def test_is_qwc(self):
...
+                qml.Identity(1) @ qml.PauliY(2),
+                qml.Identity(0),
+            ],
+            True,
+        ),  # multi I
+        ([qml.PauliZ(0) @ qml.PauliZ(1), qml.PauliZ(2), qml.PauliX(1) @ qml.PauliY(2)], False),
+    ]
+
+    @pytest.mark.parametrize("obs_lst, expected_qwc", obs_lsts)
+    def test_are_qwc_pauli_words(self, obs_lst, expected_qwc):
```
Comment (antalszava) [inline tests/grouping/test_grouping_utils.py:215]: Docstring seems to be missing.

File: doc/releases/changelog-dev.md [hunk 12]
```diff
@@ -195,6 +194,11 @@
   DeviceArray([-0.78849435, -0.8287073 , -0.85608006], dtype=float32)
   ```
 
+* Adds an `are_qwc_pauli_words(lst_pauli_words)` function which checks if a list of 
+  pauli words is qubit-wise commuting. This improves performance when measuring hamiltonians
```
Comment (antalszava) [inline doc/releases/changelog-dev.md:198]:
```suggestion
* Added an `are_pauli_words_qwc` function which checks if certain 
  Pauli words are pairwise qubit-wise commuting. This new function improves performance when measuring hamiltonians 
```

File: pennylane/grouping/utils.py [hunk 13]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
     return True
 
 
+def are_pauli_words_qwc(lst_pauli_words):
+    """Given a list of observables assumed to be valid Pauli words, determine if they
+     are all mutually qubit-wise commuting.
```
Comment (antalszava) [inline pennylane/grouping/utils.py:663]:
```suggestion
     are pairwise qubit-wise commuting.
```

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @Jaybsoni, the overall logic is looking great! :tada: 

Left a couple of minor comments and suggestions.

File: pennylane/grouping/transformations.py [hunk 14]
```diff
@@ -137,39 +135,9 @@ def diagonalize_qwc_pauli_words(qwc_grouping):  # pylint: disable=too-many-branc
...
-                if not is_qwc(
-                    pauli_binaries[i],
-                    pauli_binaries[j],
-                ):
-                    raise ValueError(
-                        f"{qwc_grouping[i]} and {qwc_grouping[j]} are not qubit-wise commuting."
-                    )
+
+    if not are_pauli_words_qwc(qwc_grouping):
+        raise ValueError("the list of pauli-words are not qubit-wise commuting.")
```
[the remark this answers is not in the document]
Comment (Jaybsoni) [reply pennylane/grouping/transformations.py:140]: I checked the documentation and it usually written like Pauli words. I will standardize that 👍🏼

Comment (Jaybsoni) [inline pennylane/grouping/transformations.py:140]:
```suggestion
        raise ValueError("The list of Pauli words are not qubit-wise commuting.")
```

File: pennylane/grouping/utils.py (hunk 6 above)
Comment (Jaybsoni) [reply pennylane/grouping/utils.py:665]: $n$ in this case refers to the maximum number of wires that must be used to describe every observable in the list. 

eg. ```obs_lst = [qml.PauliX(0), qml.PauliX(1)@qml.PauliZ(0), qml.Identity(3), qml.PauliY(2)] ```
$n = 4$ as we would need a total of 4 wires on the device to support all of these observables.

Comment (Jaybsoni) [inline pennylane/grouping/utils.py:665]:
```suggestion
    This implementation has time complexity ~ O(m * n) for m Pauli words and n wires.
    Where the number of wires is just the number of distinct wires used to describe all of the observables.  
```

File: pennylane/grouping/utils.py (hunk 8 above)
Comment (Jaybsoni) [inline pennylane/grouping/utils.py:679]:
```suggestion
        for name, wire in zip(op_names, op_wires):  # iterate over wires of the observable,
```

File: pennylane/grouping/utils.py (hunk 6 above)
Comment (Jaybsoni) [reply pennylane/grouping/utils.py:665]: How does this sound @antalszava ?

File: pennylane/grouping/utils.py (hunk 9 above)
Comment (Jaybsoni) [reply pennylane/grouping/utils.py:673]: Added!

File: pennylane/grouping/utils.py (hunk 8 above)
Comment (Jaybsoni) [reply pennylane/grouping/utils.py]: Added!

File: tests/grouping/test_grouping_utils.py (hunk 11 above)
Comment (Jaybsoni) [reply tests/grouping/test_grouping_utils.py:215]: added!

File: pennylane/grouping/utils.py (hunk 9 above)
Comment (Jaybsoni) [reply pennylane/grouping/utils.py]: Great catch on the code readability there, I would have totally missed it. But looking back this is much nicer

File: pennylane/grouping/utils.py (hunk 6 above)
Comment (antalszava) [reply pennylane/grouping/utils.py:665]: >This implementation has time complexity ~ O(m * n) for m Pauli words and n wires, where n is the number of distinct wire labels used to represent the Pauli words.

File: tests/grouping/test_grouping_utils.py [hunk 15]
```diff
@@ -195,6 +196,28 @@ def test_is_qwc(self):
...
+        """Given a list of Pauli words test that this function accurately
+        determines if they are pairwise, qubit-wise commuting."""
```
Comment (antalszava) [inline tests/grouping/test_grouping_utils.py:217]:
```suggestion
        determines if they are pairwise qubit-wise commuting."""
```

**Review by antalszava — APPROVED** [review]
Looks good :+1: :slightly_smiling_face:

File: pennylane/grouping/utils.py [hunk 16]
```diff
@@ -658,6 +658,38 @@ def is_qwc(pauli_vec_1, pauli_vec_2):
     return True
 
 
+def are_pauli_words_qwc(lst_pauli_words):
+    """Given a list of observables assumed to be valid Pauli words, determine if they
+     are pairwise qubit-wise commuting.
+
+    This implementation has time complexity ~ O(m * n) for n wires and m Pauli words.
```
Comment (Jaybsoni) [inline pennylane/grouping/utils.py:665]:
```suggestion
    """Given a list of observables assumed to be valid Pauli words, determine if they
     are pairwise qubit-wise commuting.

    This implementation has time complexity ~ O(m * n) for m Pauli words and n wires, where n is the number of distinct wire labels used to represent the Pauli words.
```

**Pull request merged**